File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/python/a_websocket_server_1.py
# FastAPI / Starlette expose WebSockets natively (the `websockets` lib is an alternative).
from fastapi import FastAPI, WebSocket, WebSocketDisconnect

app = FastAPI()

@app.websocket("/ws")
async def ws_handler(ws: WebSocket):
    # Validate origin yourself before accepting (sec 14); then complete the handshake.
    await ws.accept()                      # performs the 101 handshake (sec 4)
    try:
        while True:                        # the READ LOOP ,  one coroutine per connection
            msg = await ws.receive_text()  # awaits the next message
            await ws.send_text(msg)        # echo it straight back
    except WebSocketDisconnect:
        pass                               # client closed or connection died (sec 6)
    # FastAPI cleans up the connection when the coroutine returns

# uvicorn app:app  ,  uvicorn speaks the WebSocket protocol for you

File: 24.Web-Sockets-And-Real-time-Communication-with-WebSockets/code/go/a_websocket_server_1.go
// github.com/gorilla/websocket
var upgrader = websocket.Upgrader{
    // CheckOrigin guards against cross-site hijacking ,  DO NOT leave it open (sec 14).
    CheckOrigin: func(r *http.Request) bool {
        return r.Header.Get("Origin") == "https://app.example.com"
    },
}

func wsHandler(w http.ResponseWriter, r *http.Request) {
    conn, err := upgrader.Upgrade(w, r, nil) // performs the 101 handshake (sec 4)
    if err != nil {
        return // upgrader already wrote an HTTP error
    }
    defer conn.Close() // always close -> frees the socket (sec 19)

    for { // the READ LOOP ,  one per connection
        mt, msg, err := conn.ReadMessage()
        if err != nil {
            break // client closed or connection died (sec 6) -> exit, cleanup via defer
        }
        // echo it straight back
        if err := conn.WriteMessage(mt, msg); err != nil {
            break
        }
    }
}

func main() {
    http.HandleFunc("/ws", wsHandler) // a normal HTTP route that upgrades
    http.ListenAndServe(":8080", nil)
}
